[FLINK-10599][Documentation] Provide documentation for the modern kafka connector (#6889)
apache/flink: docs/dev/connectors/kafka.md

        <td>0.11.x</td>
        <td>Since 0.11.x Kafka does not support scala 2.10. This connector supports <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging">Kafka transactional messaging</a> to provide exactly once semantic for the producer.</td>
    </tr>
    <tr>
        <td>flink-connector-kafka_2.11</td>
        <td>1.7.0</td>
        <td>FlinkKafkaConsumer<br>
        FlinkKafkaProducer</td>

>         <td>>= 1.x</td>

Reviewer: `>= 1.0.0`

>         <td>1.x,2.x</td>

Reviewer 2: ```suggestion
        <td>>= 1.x</td>
```

* Follow the instructions from [Kafka's quickstart](https://kafka.apache.org/documentation.html#quickstart) to download the code and launch a server (launching a Zookeeper and a Kafka server is required every time before starting the application).
* If the Kafka and Zookeeper servers are running on a remote machine, then the `advertised.host.name` setting in the `config/server.properties` file must be set to the machine's IP address.

## Modern Kafka Connector

Starting with Flink 1.7, there is a new Kafka connector that does not track a specific Kafka major version. Rather, it tracks the latest version of Kafka at the time of the Flink release.

If your Kafka broker version is 1.0.0 or newer, you should use this Kafka connector. If you use an older version of Kafka (0.11, 0.10, 0.9, or 0.8), you should use the connector corresponding to the broker version.

### Compatibility

> The modern Kafka connector is compatible with older and newer Kafka brokers through the compatibility guarantees of the Kafka client API and broker. The modern Kafka client is compatible with broker versions 0.10.0 or later, depending on the features used. For details on Kafka compatibility, please refer to the [Kafka documentation](https://kafka.apache.org/protocol.html#protocol_compatibility).

Reviewer: Have we tested that this new connector works with Kafka 0.10? Definitely not in `exacty_once` mode. Maybe it does work with `at_least_once` mode, but if we haven't tested that, I would refrain from saying it. So I would rephrase this to:

```
The modern Kafka connector is compatible with broker versions 0.11.0 or later
```

> The the modern Kafka connector is compatible with older and newer Kafka brokers through the compatibility guarantees of the Kafka client API and broker. The modern Kafka client is compatible with broker versions 0.10.0 or later, depending on the features used. Fo details on Kafka compatibility, please refer to the [Kafka documentation](https://kafka.apache.org/protocol.html#protocol_compatibility).

Reviewer 3: double "the"; typo: Fo -> For

### Usage

The use of the modern Kafka connector add a dependency to it:

{% highlight xml %}
<dependency>
  <groupId>org.apache.flink</groupId>
  <artifactId>flink-connector-kafka{{ site.scala_version_suffix }}</artifactId>

>   <version>{{site.version }}</version>

Reviewer 3: sync number of spaces around `site.version`

</dependency>
{% endhighlight %}

Then instantiate the new source (`FlinkKafkaConsumer`) and sink (`FlinkKafkaProducer`). The API is the backwards compatible with the older Kafka connectors.

## Kafka Consumer

> Flink's Kafka consumer is called `FlinkKafkaConsumer08` (or `09` for Kafka 0.9.0.x versions, etc.). It provides access to one or more Kafka topics.

Reviewer: Should this line also be adjusted? I think we should also mention here that "modern" connector brakes the naming convention.

Same applies to similar sentence later for the producer:
```
Flink’s Kafka Producer is called FlinkKafkaProducer011 (or 010 for Kafka 0.10.0.x versions, etc.). It allows writing a stream of records to one or more Kafka topics.
```

Also in `Kafka Producers and Fault Tolerance` section, rename the title `Kafka 0.11` to `Kafka 0.11 and newer`, and also adopt the content that section (the name `FlinkKafkaProducer011 ` is used there frequently and with modern connector it should also be adjusted).

* Follow the instructions from [Kafka's quickstart](https://kafka.apache.org/documentation.html#quickstart) to download the code and launch a server (launching a Zookeeper and a Kafka server is required every time before starting the application).
* If the Kafka and Zookeeper servers are running on a remote machine, then the `advertised.host.name` setting in the `config/server.properties` file must be set to the machine's IP address.

> ## Modern Kafka Connector

Reviewer 2: I am wondering if instead of "Modern" we should just call it "unversioned".

Author: I don't object, what do you think? cc @aljoscha

Reviewer: Both of those names can be confusing for the user. Maybe we should name this section as `Kafka 0.11+ Connector`?

Starting with Flink 1.7, it will provide a new connector for the modern Kafka. It is different from any previous Kafka connector, it does not bind a fixed version of Kafka client, but it will track and support the latest version of Kafka Client.
The modern Kafka connector does not rely on any older version of the kafka connector and relies only on the underlying implementation of the kafka connector, which allows it to get rid of the historical burden and keep up with the latest version of the Kafka client. 

> Note that if your version of Kafka Server exceeds 0.11, then we recommend you use the modern Kafka connector. The specific version of the kafka connector will continue to maintain and fix related issues. When you use the old version of Kafka (for example, 0.11, 0.10, 0.9, 0.8), we still recommend that you use the connector corresponding to the corresponding version, because they are precisely matched.

Reviewer 2: ```suggestion
If your Kafka broker version is 1.0.0 or newer, you should use this Kafka connector. If you use an older version of Kafka (0.11, 0.10, 0.9, or 0.8), you should use the connector corresponding to the broker version.
```

### Compatibility

> The compatibility of the modern Kafka connector is completely dependent on the backward compatibility guarantee of the Kafka Client API (Kafka supports the latest Client API to access the old version of Kafka Server. For more details on Kafka compatibility, please refer to the [official Kafka documentation](https://kafka.apache.org/protocol.html#protocol_compatibility)). Therefore, as long as the connector is upgraded and supported to the latest Kafka Client version, it can provide the same compatibility guarantee as the Kafka Client.

Reviewer 2: ```suggestion
The the modern Kafka connector is compatible with older and newer Kafka brokers through the compatibility guarantees of the Kafka client API and broker. The modern Kafka client is compatible with broker versions 0.10.0 or later, depending on the features used. Fo details on Kafka compatibility, please refer to the [Kafka documentation](https://kafka.apache.org/protocol.html#protocol_compatibility).
```

### Usage

> The use of the modern Kafka connector is very simple. We only need to introduce dependencies on it:

Reviewer 2: ```suggestion
The use of the modern Kafka connector add a dependency to it:
```

{% highlight xml %}
<dependency>
  <groupId>org.apache.flink</groupId>
  <artifactId>flink-connector-kafka{{ site.scala_version_suffix }}</artifactId>
  <version>{{site.version }}</version>
</dependency>
{% endhighlight %}

> Then instantiate the new source (`FlinkKafkaConsumer`) and sink (`FlinkKafkaProducer`). As for the API, it is fully compatible with other Kafka connectors that already exist. So in theory, you don't need to make any changes to it to make your program work.

Reviewer 2: ```suggestion
Then instantiate the new source (`FlinkKafkaConsumer`) and sink (`FlinkKafkaProducer`). The API is the backwards compatible with the older Kafka connectors.
```
